clang/lib/Sema/SemaSYCL.cpp

```diff
@@ -4570,6 +5660,33 @@ void Sema::MarkDevices() {
   }
 }
 
+void Sema::ProcessFreeFunctions() {
+  for (Decl *D : syclDeviceDecls()) {
+    auto *SYCLKernel = cast<FunctionDecl>(D);
+    if (IsFreeFunction(*this, SYCLKernel)) {
+      SyclKernelDecompMarker DecompMarker(*this);
```

Comment:
Is there a reason you did not handle the diagnostics in `SemaChecking`, like we currently do so for `CheckSYCLKernelCall`